crates/client/rpc/src/lib.rs

```diff
@@ -195,6 +199,25 @@ where
             None
         })
     }
+
+    /// Returns the state diff for the given block.
+    ///
+    /// # Arguments
+    ///
+    /// * `block_hash` - The hash of the block containing the state diff (starknet block).
+    fn get_state_diff(&self, block_hash: B::Hash, starknet_block_hash: &BlockHash) -> Result<StateDiff, String> {
```

Comment:
There are only 2 places we read `state_diff` right now. One is here and the other is when doing the state update. The conversion from `ThinStateDiff` to `Vec<U256>` also happens during state update so we already have the vector form available in memory over there. So why don't we just store actual state diff which we need to read here, that will save us the decoding part?

Moreover, if the update_state part would anyways move to a new service, then this is the only place where we read it right?

Reply:
Unfortunately, we can't truly rely on this. I think ideally, we'd want to have a fallback mechanism that fetches the state diff from L1 for instance. We might also have concurrency issues with this approach. 

Your idea is to directly store the `starknet_rs::StateDiff` in the DA db ? And completely get rid of `ThinStateDiff` ?

Reply:
You mean we can't rely on an external service doing DA or we can't rely on storing state diffs directly?

I mean we can get rid of the decode function if we just store `StateDiff` directly. There's no need to encode it, store and then decode it. The only place I see the encoded value being used is during update state and that's also where we do the encoding, so its already in memory.

Reply:
We could store `ThinStateDiff` or `StateDiff` whatever is convenient.

Reply:
Unfortunately, neither of them is SCALE encodable, what's the best approach there ? I think it ends up the same overhead as encoding/decoding to Vec<U256> no?

Reply:
And here we decode the whole thing when we only need a subpart of it

Reply:
SCALE could easily be impl on starknet_api ThinStateDiff and StateDiff, they both use IndexMap now

Reply:
What about the part where we store the whole `BlockDAData` instead of just the state diff. It does not sound like a huge issue but wdyt?

Reply:
So what is happening right now is:
ThinStateDiff -> Vec<U256> -> SCALE encode -> SCALE decode -> StateDiff

What you think would be better is:
ThinStateDiff -> SCALE encode -> SCALE decode -> Convert to StateDiff

What I'm saying is that if we can avoid this conversion function between `ThinStateDIff` and `StateDiff` that would be better and this `Vec<U256>` intermediate representation is perfect for this.

Reply:
But why is a converstion b/w `ThinStateDIff` and `StateDiff` more difficult, I think it will be more readable.

Reply:
also, ya we can rename the function `state_diff_to_calldata` to `block_data_to_call_data`?

Reply:
> But why is a converstion b/w `ThinStateDIff` and `StateDiff` more difficult, I think it will be more readable.

Maybe more readable and even that I'm not sure but high probably less efficient if we have to loop over all the mappings and convert the types.

Reply:
But we are anyways looping over all the mappings and converting, we just have an extra intermediate state.